File: scripts/generate-icons.py
"""Build build/icon.icns from a square PNG source.

The macOS path (scripts/generate-icons.mjs) shells out to sips and iconutil.
Neither exists off macOS, so this writes the ICNS container directly: it emits
the same nine chunks iconutil does (TOC plus ic07-ic14), each holding an
optimised PNG.
"""
from __future__ import annotations

import struct
import sys
from io import BytesIO
from pathlib import Path

from PIL import Image

# (chunk type, pixel size) — mirrors what `iconutil -c icns` produces.
CHUNKS = [
    (b"ic07", 128),
    (b"ic08", 256),
    (b"ic09", 512),
    (b"ic10", 1024),
    (b"ic11", 32),
    (b"ic12", 64),
    (b"ic13", 256),
    (b"ic14", 512),
]


def _png(source: Image.Image, size: int) -> bytes:
    resized = source.resize((size, size), Image.LANCZOS)
    buffer = BytesIO()
    resized.save(buffer, format="PNG", optimize=True, compress_level=9)
    return buffer.getvalue()


def build_icns(source_path: Path, target_path: Path) -> None:
    source = Image.open(source_path).convert("RGBA")
    if source.width != source.height:
        raise SystemExit(f"Icon source must be square, got {source.width}x{source.height}")

    payloads = [(kind, _png(source, size)) for kind, size in CHUNKS]

    # The table of contents lists every following chunk's type and total length.
    toc_body = b"".join(kind + struct.pack(">I", len(data) + 8) for kind, data in payloads)
    toc = b"TOC " + struct.pack(">I", len(toc_body) + 8) + toc_body
    body = toc + b"".join(
        kind + struct.pack(">I", len(data) + 8) + data for kind, data in payloads
    )

    target_path.write_bytes(b"icns" + struct.pack(">I", len(body) + 8) + body)


def main() -> None:
    build_dir = Path(__file__).resolve().parent.parent / "build"
    source_path = Path(sys.argv[1]).resolve() if len(sys.argv) > 1 else build_dir / "icon.png"
    if not source_path.exists():
        raise SystemExit(f"Icon source not found: {source_path}")

    icns_path = build_dir / "icon.icns"
    build_icns(source_path, icns_path)
    print(f"Generated {icns_path} ({icns_path.stat().st_size // 1024} KB)")


if __name__ == "__main__":
    main()

File: scripts/generate-icons.mjs
import { execFileSync } from "node:child_process";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const rootDir = path.resolve(__dirname, "..");
const buildDir = path.join(rootDir, "build");
const sourceArg = process.argv[2];
const defaultSource = path.join(buildDir, "icon.png");
const sourcePath = path.resolve(sourceArg ?? defaultSource);

if (!fs.existsSync(sourcePath)) {
  console.error(`Icon source not found: ${sourcePath}`);
  process.exit(1);
}

// sips and iconutil ship with macOS and exist nowhere else, so off-macOS fall
// back to the Pillow writer, which emits the same ic07-ic14 chunk set.
function hasMacIconTools() {
  try {
    execFileSync("which", ["sips", "iconutil"], { stdio: "ignore" });
    return true;
  } catch {
    return false;
  }
}

if (!hasMacIconTools()) {
  execFileSync("python3", [path.join(__dirname, "generate-icons.py"), sourcePath], {
    stdio: "inherit"
  });
  process.exit(0);
}

const iconPngPath = path.join(buildDir, "icon.png");
execFileSync("sips", ["-s", "format", "png", sourcePath, "--out", iconPngPath, "-z", "1024", "1024"], {
  stdio: "inherit"
});

const iconsetDir = path.join(buildDir, "icon.iconset");
fs.rmSync(iconsetDir, { recursive: true, force: true });
fs.mkdirSync(iconsetDir, { recursive: true });

for (const size of [16, 32, 128, 256, 512]) {
  const double = size * 2;
  execFileSync(
    "sips",
    ["-z", String(size), String(size), iconPngPath, "--out", path.join(iconsetDir, `icon_${size}x${size}.png`)],
    { stdio: "ignore" }
  );
  execFileSync(
    "sips",
    [
      "-z",
      String(double),
      String(double),
      iconPngPath,
      "--out",
      path.join(iconsetDir, `icon_${size}x${size}@2x.png`)
    ],
    { stdio: "ignore" }
  );
}

execFileSync("iconutil", ["-c", "icns", iconsetDir, "-o", path.join(buildDir, "icon.icns")], {
  stdio: "inherit"
});
fs.rmSync(iconsetDir, { recursive: true, force: true });

console.log(`Generated ${path.join(buildDir, "icon.png")}`);
console.log(`Generated ${path.join(buildDir, "icon.icns")}`);
